block: expose io_engine in the API and add Async option (#2787)
firecracker-microvm/firecracker: docs/api_requests/block-io-engine.md

executing requests in an async manner, therefore getting overall higher
throughput by taking better advantage of the block device hardware which
typically supports queue depths greater than 1.

The block IO engine is configured via the PUT /drives API call (pre-boot only),
with the `io_engine` field taking two possible values:

- `Sync` (default)
- `Async`

The `Sync` variant is the default, in order to provide backwards compatibility
with older Firecracker versions.

## Example configuration

```bash
curl --unix-socket ${socket} -i \
     -X PUT "http://localhost/drives/rootfs" \
     -H "accept: application/json" \
     -H "Content-Type: application/json" \
     -d "{
             \"drive_id\": \"rootfs\",
             \"path_on_host\": \"${drive_path}\",
             \"is_root_device\": true,
             \"is_read_only\": false,
             \"io_engine\": \"Async\"
         }"
```

## Host requirements

Firecracker requires a minimum host kernel version of 5.10.51 for the `Async`
IO engine.

This requirement is based on the availability of the `io_uring` subsystem, as
well as a couple of features and bugfixes that were added in newer kernel

> versions.

Reviewer: Maybe some recommendations when to use one or the other.
Also you should mention what happens when one uses io_uring with kernel version lower than 5.10.

Author: I added a couple of recommendations, which we will have to revisit once we complete the snapshotting support. I'd also like to get Serban's input on them.
I also mentioned what happens when you try to use io_uring on a kernel older than 5.10

# Block device IO engine

> Prior to Firecracker 1.0.0, the emulated block device uses a synchronous IO

Reviewer 2: ```suggestion
Prior to Firecracker 1.0.0, the emulated block device used a synchronous IO
```

Reviewer 2: I get your reasoning, but it just sounds a bit weird to me to keep present tense here. How about rephrasing to something like:

```suggestion
For all Firecracker versions prior to v1.0.0, the emulated block device uses a synchronous IO 
engine for executing the device requests, based on blocking system calls for.
```
so the it would be clear that firecracker supports Sync for all versions and the next paragraph specifies that Async is only available starting with Firecracker v1.0.0.

Author: done 👍🏻

engine for executing the device requests, based on blocking system calls.

> Firecracker 1.0.0 adds support for a new, asynchronous block device IO engine.

Reviewer 2: I wouldn't add a comma here.
```suggestion
Firecracker 1.0.0 adds support for a new asynchronous block device IO engine.
```

The `Async` engine leverages [`io_uring`](https://kernel.dk/io_uring.pdf) for
executing requests in an async manner, therefore getting overall higher

> throughput by taking better advantage of the block device hardware which

Reviewer 2: nit:
```suggestion
throughput by taking better advantage of the block device hardware, which
```
